### ISSUE TITLE
Reduced performance of sending logs

### ISSUE THREAD

**Issue by KhafRuslan** [metrico/otel-collector#92]
At a certain point, when we reached a heavy load we encountered the problem of low speed of sending logs via promtail 
![image](https://github.com/user-attachments/assets/129bf2ee-1647-499d-bd6c-2784a521333d)
The difference is the speed of reading promtail logs from a file with the same configuration. In the screenshot promtail sent all messages to loki
![image](https://github.com/user-attachments/assets/2dfcc0db-7b40-4429-a574-736a8cb2bb9b)

Configuring the client part of promtail:
`
clients:
  - url: http://127.0.0.1:3111/loki/api/v1/push
    batchwait: 1s
    batchsize: 100
    backoff_config:
      min_period: 100ms
      max_period: 5s
    external_labels:
      job: ${HOSTNAME}
`
The solution was simple, we raised the second loki log receiver. After that we can observe a decrease in the graph above. The result is the same 
![image](https://github.com/user-attachments/assets/14379ba6-4522-4387-8c45-a305d06ee208)
The average resource utilization of an instance was no higher than 30 percent

**Comment by lmangani** [comment]
The qryn process is single threaded so you either need to scale multiple writers/readers and distribute traffic to achieve your desired capacity or use the [qryn otel-collector](https://github.com/metrico/otel-collector/i) and write directly into ClickHouse at max speed.

**Comment by KhafRuslan** [comment]
Rather, the description of the panels was confusing. I use qryn otel-collector, it was on it that I encountered the problem  
```
  loki:
    protocols:
       grpc:
        endpoint: 0.0.0.0:3200
       http:
        endpoint: 0.0.0.0:3100
```

**Comment by lmangani** [comment]
If you are using the otel-collector to ingest, then I would assume the bottleneck being either with the collector or clickhouse rather than qryn itself. Did you observe any resource bottlenecking while operating the setup?

**Comment by KhafRuslan** [comment]
No, I did not observe such a problem, it resolved immediately after I started sending to another receiver

**Comment by lmangani** [comment]
> I'm not sure if it's a resource problem, because it works correctly when I bring up another receiver

We definitely need to investigate this further to understand what the root cause is. Could you show the multi-receiver config too?

**Comment by rita7lopes** [comment]
Hello @KhafRuslan,

Could you retest on the latest image or the collector and gigapipe? 
Nothing in the log path has changed specifically to address ingest throughput, so this likely still reproduces, but we'd like to confirm against current before investigating. 

If it persists, please share collector CPU/mem and the sending_queue/num_consumers settings alongside the promtail config.

Thanks for reporting!

**Comment by rita7lopes** [comment]
Adding some analysis while we wait on the retest, to narrow this down.

The key clue is that adding a **second receiver** fixed it, rather than any exporter tuning. That points at the **receive/transport path**, not the ClickHouse write path: a single promtail client (`batchsize: 100`, `batchwait: 1s`, one connection) serializes its pushes over one connection into a single receiver's HTTP handler, and a second receiver just splits that load. If the ceiling were the exporter, raising `sending_queue.num_consumers` would have had the same effect as adding a receiver.

So, in addition to the retest, two things would help a lot:

1. Before adding receivers, try raising `sending_queue.num_consumers` (and `queue_size`) on a single pipeline and see whether it then keeps up — that isolates exporter parallelism from a receiver-side limit.
2. If you can, grab a **CPU profile** via the `pprof` extension during the slow period. That's the fastest way to tell a receiver/transport bottleneck apart from exporter-side CPU or ClickHouse write latency.

On our side: while reviewing the log exporter we found an unrelated efficiency issue in its hot path (a redundant per-record resource copy) that we'll fix separately. It likely isn't the root cause here — the second-receiver behaviour above suggests the limit is upstream of the exporter — but a profile would confirm whether it contributes.
